src/main.rs

```diff
@@ -299,6 +310,28 @@ fn check_vector_width() {
     CSR.sstatus.read_and_clear_bits(sstatus::vs::Dirty.value);
 }
 
+/// The entry point for the test runner
+#[cfg(test)]
+#[no_mangle]
+extern "C" fn test_runner_kernel(_hart_id: u64, _fdt_addr: u64) {
```

Comment:
just call this `kernel_init` and put a `#[cfg(not(test))]` over the other `kernel_init`.

Reply:
in that case, maybe we don't need another start.S

Will fix